### PR TITLE
Formfiller

### LINKED ISSUE #197: [0.9] Repair form filler
Upon hitting `zn` an empty (!) text file is opened in the editor.

Upon saving, a file is written (eg. data/dforms/duckduckgo.com.default).

Upon hitting `ze` again an empty file is opened which will overwrite the original one upon saving.

`zl` stays without effect.

If I remember correctly the file used to have the field names or id's pre-filled, but even then, formfiller "load" never worked for me.

### PR DESCRIPTION
**Pull request by keis**
Various fixes to make the formfiller work again. Fixes #197

### CODE REVIEW

File: src/commands.c [hunk 1]
```diff
@@ -511,6 +513,22 @@ split_quoted (const gchar *src, const gboolean unquote)
     return argv;
 }
 
+static gchar *
+unescape (gchar *src) {
+    gchar *s = src, *p = src;
```
Comment (mathstuf) [inline src/commands.c]: Never did like this style of declarations. Two lines make the pointer thing clearer. Also, lone brace line for functions.

File: src/io.c [hunk 2]
```diff
@@ -685,37 +685,39 @@ send_buffered_event (gpointer event, gpointer data)
...
 
     static const gsize BUFSZ = 256;
     gchar input[BUFSZ];
     gsize read;
     GError *error = NULL;
-    gboolean eof = FALSE;
 
-    do {
-        GIOStatus status = g_io_channel_read_chars (gio, input, BUFSZ, &read, &error);
+    if (condition > 10) {
```
Comment (mathstuf) [inline src/io.c]: Debugging? What is 10?

Comment (keis) [inline src/io.c]: ah yes. just debug stuff from trying to figure out what the conditions I got back was. Should be removed

**Comment by GSI** [comment]
Unless you're already at it, I'd like to help DRY up **formfiller.sh** and add some small improvements to it's logic.

Things I notice from a first quick look:

* `get_option` returns absolute paths, but is expected to return just the profile suffix by it's callers.
* `ls` output is parsed, which should be avoided.
* Some possibility to improve code readability. Question regarding line 60: What's the idea of counting by making the string longer? Faster than `expr`?

**Comment by mathstuf** [comment]
The first two bullets are already handled by this PR. The last one is probably because `$(( ))` isn't allowed and ignorance of `expr` :) .

**Comment by keis** [comment]
@GSI I'm happy with just getting in to state where it works again. Feel free to improve on and refactor the script :)

File: src/io.c (hunk 2 above)
Comment (keis) [inline src/io.c]: Fixed

**Comment by GSI** [comment]
Confirming working formfiller functionality with this PR applied;

I notice a random delay of <= 3 seconds upon `load` (`zl`).

Also seeing these error messages (right AFTER `load_profile` function finishes):

```
** Message: console message: /usr/share/uzbl/examples/data/scripts/formfiller.js @90: Error occurred when applying formfiller text: TypeError: Attempted to assign to readonly property.

** Message: console message: /usr/share/uzbl/examples/data/scripts/formfiller.js @90: Error occurred when applying formfiller text: TypeError: Attempted to assign to readonly property.
```

**Comment by GSI** [comment]
@mathstuf:
> The first two bullets are already handled by this PR.

The first, yes. The second means that we should avoid `ls`.
(http://mywiki.wooledge.org/ParsingLs ; will refactor accordingly once this PR is merged.)

**Comment by mathstuf** [comment]
Oh, I see the remaining `ls` that is left. I guess `echo` would be just as good really since we're really relying on the shell to do the glob expansion…

**Comment by keis** [comment]
Trying to set read-only properties is no big deal it just fails as expected. We should perhaps filter those in `.dump()`

I don't see a delay when trying it here so I think that might be related to some buffering/delay in uzbl overall. Maybe depending on kernel config or X versions etc.

**Comment by keis** [comment]
Fixed more shell issues in `once` and `edit` (with no profile). Making a note of the read-only issue in #207. 

I changed everything to use full paths mostly because that was what the majority of the code was using already. If you end up doing a bigger clean up feel free to change that to whatever makes more sense.

**Pull request merged**